### ISSUE TITLE
Question: How should I expect to handle passing the io object around?

### ISSUE THREAD

**Issue by ajfisher** [monteslu/udp-serial#1]
So first off is that this is all working fine with johnny five - it basically "just works" out of the box if you pass the IO object in on the board constructor.

I'm trying to get node-pixel to work with this as well which is obviously not a johnny five component but it consumes a board object or a firmata object to then pipe it's writes etc along that.

So basically if I do this:

```
strip = new pixel.Strip({
             data: 6,
             length: 8,
             firmata: this,
             controller: "FIRMATA",
});
```

Where `this` in this case is the `board` it should in node-pixel detect that the board has an io object and therefore just run firmata over it. Unfortuantely I'm getting undefined errors here. 

If I pass the IO object specifically then I'm getting errors in my firmata.sp.write() methods as there's no sp.write as there's no _real_ serial port.

I think I'm missing a bit of knowledge around the way the virtual serial port works in this case and how the object maps down onto the board and so as a result I'm either not passing it properly to node-pixel or my node-pixel firmata implementation isn't generous enough in the objects it expects to be handed that "look" like firmata.

**Comment by monteslu** [comment]
Do you have the latest firmata.js?  There was some recent juggling of the sp/transport properties, but I think it should be good now.

Other than that, maybe just need a second set of eyes on how you're setting the whole thing up.  Possible this context changed?

**Comment by monteslu** [comment]
This gives me a function:

``` javascript

'use strict';

var five = require('johnny-five');
var firmata = require('firmata');

var SerialPort = require('udp-serial').SerialPort;

var sp = new SerialPort({host:'localhost', port:4567});

var io = new firmata.Board(sp);

var board = new five.Board({io: io});

console.log(board.io.sp.write);

```

**Comment by ajfisher** [comment]
So this is where it appears to be failing:

https://github.com/ajfisher/node-pixel/blob/master/lib/pixel.js#L93-L101

In this case I'm assuming a firmata instance that's either real firmata or an io instance and that it should have a writable sp associated to it.

Is this an incorrect assumption?

**Comment by monteslu** [comment]
That is correct.  It should be working.   Nothing is jumping out to me as to why it isn't.

**Comment by ajfisher** [comment]
I think this might have been to do with the way the port was being selected
in firmata. As it happens I've just refactored a bunch of this code in
order to get node pixel to work with testing and I think it might work now.
Will give it a go tonight.
